src/System.Buffers/src/System/Buffers/BufferInternal.cs

```diff
@@ -192,5 +192,118 @@ public static unsafe void MemoryCopy(void* source, void* destination, int destin
 
             Memmove((byte*)destination, (byte*)source, checked((uint)sourceBytesToCopy));
         }
+        
+        // TODO: This is naive implementation. Do we have any better native implementation?
+        // TODO: Should we start thinking about supporting long lengths?
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe bool MemoryEqual(void* lhs, int lhsLength, void* rhs, int rhsLength)
+        {
+            // 1. Fast paths
+            if (lhsLength != rhsLength)
+                return false;
+            if (lhs == rhs)
+                return true;
+
+            switch (lhsLength)
+            {
+                case 0: return true;
+                case 1: return *(byte*)lhs == *(byte*)rhs;
+                case 2: return *(UInt16*)lhs == *(UInt16*)rhs;
+                case 3: return *(UInt16*)lhs == *(UInt16*)rhs &&  *((byte*)lhs + 2) == *((byte*)rhs + 2);
+                case 4: return *(UInt32*)lhs == *(UInt32*)rhs;
+                case 8: return *(UInt64*)lhs == *(UInt64*)rhs;
+            }
+            
+            // 2. We split bytes into two groups of: (n/4)*4 and n%4 bytes
+            
+            // TODO: should we play with alignment here?
```

Comment:
I was actually thinking about doing that but then thought about the case where one buffer is aligned and the other one isn't and wasn't sure how to handle. I think we should rather move to some known and well perf tested implementation (maybe there is something like that in one of the C++ standard libraries)